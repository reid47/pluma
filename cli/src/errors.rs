use std::fmt;

#[derive(Debug)]
pub enum UsageError {
  NoCommand,
  UnknownCommand(String),
  MissingEntryPath,
  InvalidEntryPath(String),
  EntryDirDoesNotContainEntryFile(String),
}

impl fmt::Display for UsageError {
  fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
    write!(f, "{:#?}", self)
  }
}