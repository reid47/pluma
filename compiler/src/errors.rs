use std::collections::HashMap;
use crate::ast::Node;

#[derive(Debug)]
pub struct PackageCompilationErrorSummary {
  pub package_errors: Vec<String>,
  pub module_errors: HashMap<String, Vec<ModuleCompilationErrorDetail>>
}

#[derive(Debug)]
pub struct ModuleCompilationErrorDetail {
  pub location: Option<(usize, usize)>,
  pub message: String,
}

#[derive(Debug)]
pub enum PackageCompilationError {
  ModulesFailedToCompile(Vec<String>),
  CyclicalDependency(Vec<String>),
}

#[derive(Debug)]
pub enum ModuleCompilationError {
  FileError(FileError),
  TokenizeError(TokenizeError),
  ParseError(ParseError),
}

#[derive(Debug)]
pub enum FileError {
  FailedToReadFile(String),
}

#[derive(Debug)]
pub enum TokenizeError {
  InvalidBinaryDigitError(usize, usize),
  InvalidHexDigitError(usize, usize),
  InvalidOctalDigitError(usize, usize),
  UnclosedStringError(usize, usize),
  UnclosedInterpolationError(usize, usize),
}

#[derive(Debug, Clone)]
pub enum ParseError {
  UnexpectedToken(usize),
  UnexpectedEOF,
  UnclosedParentheses(usize),
  UnclosedBlock(usize),
  UnclosedArray(usize),
  UnclosedDict(usize),
  UnexpectedArrayElementInDict(Node),
  UnexpectedDictEntryInArray(Node),
  UnexpectedTokenAfterDot(usize),
  UnexpectedTokenInImport(usize),
  MissingArrowInMatchCase(usize),
  MissingArrowAfterBlockParams(usize),
  MissingAliasAfterAsInImport(usize),
  MissingCasesInMatchExpression(usize),
}