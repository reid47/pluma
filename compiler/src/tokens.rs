// TODO: use this to only derive in tests
// #[cfg_attr(test, derive(Debug, PartialEq))]
#[derive(Debug, PartialEq)]
pub enum Token<'a> {
  Arrow { start: usize, end: usize },
  BinaryDigits { start: usize, end: usize, value: &'a [u8] },
  Colon { start: usize, end: usize },
  ColonEquals { start: usize, end: usize },
  Comma { start: usize, end: usize },
  Comment { start: usize, end: usize, value: &'a [u8] },
  DecimalDigits { start: usize, end: usize, value: &'a [u8] },
  Dot { start: usize, end: usize },
  DoubleArrow { start: usize, end: usize },
  DoubleColon { start: usize, end: usize },
  Equals { start: usize, end: usize },
  HexDigits { start: usize, end: usize, value: &'a [u8] },
  Identifier { start: usize, end: usize, value: &'a [u8] },
  InterpolationEnd { start: usize, end: usize },
  InterpolationStart { start: usize, end: usize },
  LeftBrace { start: usize, end: usize },
  LeftBracket { start: usize, end: usize },
  LeftParen { start: usize, end: usize },
  LineBreak { start: usize, end: usize },
  Minus { start: usize, end: usize },
  OctalDigits { start: usize, end: usize, value: &'a [u8] },
  Pipe { start: usize, end: usize },
  Plus { start: usize, end: usize },
  RightBrace { start: usize, end: usize },
  RightBracket { start: usize, end: usize },
  RightParen { start: usize, end: usize },
  StringLiteral { start: usize, end: usize, value: &'a [u8] },
  Unexpected { start: usize, end: usize },
}

pub fn get_token_location(token: &Token) -> (usize, usize) {
  match token {
    Token::Arrow { start, end } => (*start, *end),
    Token::BinaryDigits { start, end, .. } => (*start, *end),
    Token::Colon { start, end } => (*start, *end),
    Token::ColonEquals { start, end } => (*start, *end),
    Token::Comma { start, end } => (*start, *end),
    Token::Comment { start, end, .. } => (*start, *end),
    Token::DecimalDigits { start, end, .. } => (*start, *end),
    Token::Dot { start, end } => (*start, *end),
    Token::DoubleArrow { start, end } => (*start, *end),
    Token::DoubleColon { start, end } => (*start, *end),
    Token::Equals { start, end } => (*start, *end),
    Token::HexDigits { start, end, .. } => (*start, *end),
    Token::Identifier { start, end, .. } => (*start, *end),
    Token::InterpolationEnd { start, end } => (*start, *end),
    Token::InterpolationStart { start, end } => (*start, *end),
    Token::LeftBrace { start, end } => (*start, *end),
    Token::LeftBracket { start, end } => (*start, *end),
    Token::LeftParen { start, end } => (*start, *end),
    Token::LineBreak { start, end } => (*start, *end),
    Token::Minus { start, end } => (*start, *end),
    Token::OctalDigits { start, end, .. } => (*start, *end),
    Token::Pipe { start, end, .. } => (*start, *end),
    Token::Plus { start, end, .. } => (*start, *end),
    Token::RightBrace { start, end } => (*start, *end),
    Token::RightBracket { start, end } => (*start, *end),
    Token::RightParen { start, end } => (*start, *end),
    Token::StringLiteral { start, end, .. } => (*start, *end),
    Token::Unexpected { start, end } => (*start, *end),
  }
}