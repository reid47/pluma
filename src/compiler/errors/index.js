export { default as TokenizerError } from './tokenizer-error';
export { default as ParserError } from './parser-error';
