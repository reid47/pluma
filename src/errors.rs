use std::fmt;

#[derive(Debug)]
pub struct UsageError {
  message: String,
}

impl fmt::Display for UsageError {
  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
    write!(f, "{}", self.message)
  }
}

impl UsageError {
  pub fn unknown_command(command_name: String) -> UsageError {
    UsageError {
      message: format!("Unknown command: {}", command_name),
    }
  }
}
