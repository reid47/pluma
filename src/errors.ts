export class ParseError extends Error {
  constructor(line: number, col: number, message: string) {
    super(`Parse error at line ${line}, column ${col}:\n\n${message}`);

    this.stack = null;
  }
}
