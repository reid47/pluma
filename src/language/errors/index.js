export { default as TokenizerError } from './tokenizer-error';
