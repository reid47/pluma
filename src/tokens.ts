export type TokenKind =
  | 'Boolean'
  | 'Char'
  | 'Comment'
  | 'Identifier'
  | 'InterpolationEnd'
  | 'InterpolationStart'
  | 'Operator'
  | 'Number'
  | 'String'
  | '->'
  | '=>'
  | '::'
  | ':'
  | ':='
  | ','
  | '.'
  | '='
  | '{'
  | '}'
  | '['
  | ']'
  | '('
  | ')';

export interface Token {
  kind: TokenKind;
  value: string;
  lineStart: number;
  colStart: number;
  lineEnd: number;
  colEnd: number;
}
