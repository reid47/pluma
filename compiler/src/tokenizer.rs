use crate::tokens::Token;
use crate::errors::TokenizeError;

pub struct Tokenizer<'a> {
  source: &'a Vec<u8>,
  preserve_comments: bool,
  source_length: usize,
}

fn is_identifier_start_char(byte: u8) -> bool {
  (byte >= b'a' && byte <= b'z') || (byte >= b'A' && byte <= b'Z')
}

fn is_identifier_char(byte: u8) -> bool {
  (byte >= b'a' && byte <= b'z') || (byte >= b'A' && byte <= b'Z') || (byte >= b'0' && byte <= b'9')
}

fn is_digit(byte: u8) -> bool {
  byte >= b'0' && byte <= b'9'
}

impl<'a> Tokenizer<'a> {
  pub fn from_source(source: &'a Vec<u8>) -> Tokenizer<'a> {
    let length = source.len();

    return Tokenizer {
      source: source,
      preserve_comments: true,
      source_length: length,
    };
  }

  pub fn collect_tokens(&mut self) -> Result<Vec<Token<'a>>, TokenizeError> {
    let source = self.source;
    let length = self.source_length;

    let mut tokens = Vec::new();
    let mut index = 0;
    let mut line = 0;
    let mut line_start_index = 0;

    let mut string_stack = Vec::new();
    let mut interpolation_stack = 0;

    // We iterate through all chars in a single loop, appending tokens as we find them.
    // The trickiest parts here are related to string interpolations, since they can
    // be nested arbitrarily deep (e.g. "hello $("Ms. $(name)")"). These parts are
    // commented below.
    while index < length {
      let start_index = index;
      let byte = source[start_index];

      if string_stack.is_empty() && byte == b'"' {
        // If the string stack is empty and byte is ", we are at the beginning of
        // a brand new string. Save the start index and advance.
        string_stack.push((line, index - line_start_index, index));
        index += 1;
        continue;
      }

      if !string_stack.is_empty() {
        // If the string stack is not empty, we're somewhere inside a string (maybe
        // in an interpolation, though). We must check if we need to end the string,
        // start/end an interpolation, or just carry on.

        if byte == b'"' && string_stack.len() == interpolation_stack {
          // If the two stacks have the same size, we must be inside of an interpolation,
          // so the " indicates the beginning of a nested string literal. Save the index
          // in the string stack and advance.
          string_stack.push((line, index - line_start_index, index));
          index += 1;
          continue;
        }

        if byte == b'"' {
          // Here, the " must indicate the end of a string literal section. Pop from
          // the string stack, add a new token, then advance.
          let (line_start, col_start, offset) = string_stack.pop().unwrap();

          println!("str ending, {}, {}, {}", line_start, col_start, offset);

          tokens.push(Token::String {
            line_start,
            col_start,
            line_end: line,
            col_end: index + 1 - line_start_index,
            value: &source[offset + 1..index],
          });

          index += 1;
          continue;
        }

        if byte == b'$' && start_index + 1 < length && source[start_index + 1] == b'(' {
          // We must be at the beginning of an interpolation, so create a token for
          // the string literal portion leading up to the interpolation, one for the
          // interpolation start, and add to the interpolation stack.
          let (line_start, col_start, offset) = string_stack.last().unwrap();

          tokens.push(Token::String {
            line_start: *line_start,
            col_start: *col_start,
            line_end: line,
            col_end: index - line_start_index,
            value: &source[offset + 1..index],
          });

          tokens.push(Token::InterpolationStart {
            line: line,
            col: index - line_start_index,
          });

          index += 2;
          interpolation_stack += 1;
          continue;
        }

        if byte == b')' {
          // We must be at the end of an interpolation, so make a token for it and
          // fix the index on the last string in the string stack so that it starts
          // here. Decrease the interpolation stack.
          tokens.push(Token::InterpolationEnd {
            line: line,
            col: index - line_start_index,
          });

          string_stack.pop();
          string_stack.push((line, index - line_start_index + 1, index));

          index += 1;
          interpolation_stack -= 1;
          continue;
        }

        if string_stack.len() > interpolation_stack {
          // If the string stack is larger than the interpolation stack, we must be
          // inside of a string literal portion. Just advance past this char so we can
          // include it in the string later.
          index += 1;
          continue;
        }

        // At this point, we must be inside an interpolation (not a string literal),
        // so continue to collect tokens as we would outside of a string.
      }

      match byte {
        b' ' | b'\r' | b'\t' => {
          index += 1;
        }

        b'\n' => {
          index += 1;
          line += 1;
          line_start_index = index;
        }

        b'(' => {
          index += 1;

          tokens.push(Token::LeftParen {
            line: line,
            col: start_index - line_start_index,
          })
        }

        b')' => {
          index += 1;

          tokens.push(Token::RightParen {
            line: line,
            col: start_index - line_start_index,
          })
        }

        b'{' => {
          index += 1;

          tokens.push(Token::LeftBrace {
            line: line,
            col: start_index - line_start_index,
          })
        }

        b'}' => {
          index += 1;

          tokens.push(Token::RightBrace {
            line: line,
            col: start_index - line_start_index,
          })
        }

        b'[' => {
          index += 1;

          tokens.push(Token::LeftBracket {
            line: line,
            col: start_index - line_start_index,
          })
        }

        b']' => {
          index += 1;

          tokens.push(Token::RightBracket {
            line: line,
            col: start_index - line_start_index,
          })
        }

        b',' => {
          index += 1;

          tokens.push(Token::Comma {
            line: line,
            col: start_index - line_start_index,
          })
        }

        b'.' => {
          index += 1;

          tokens.push(Token::Dot {
            line: line,
            col: start_index - line_start_index,
          })
        }

        b'=' => match source.get(index + 1) {
          Some(b'>') => {
            index += 2;

            tokens.push(Token::DoubleArrow {
              line: line,
              col: start_index - line_start_index,
            })
          }

          _ => {
            index += 1;

            tokens.push(Token::Equals {
              line: line,
              col: start_index - line_start_index,
            })
          }
        },

        b'-' => match source.get(index + 1) {
          Some(b'>') => {
            index += 2;

            tokens.push(Token::Arrow {
              line: line,
              col: start_index - line_start_index,
            })
          }

          _ => {
            index += 1;

            tokens.push(Token::Minus {
              line: line,
              col: start_index - line_start_index,
            })
          }
        },

        b':' => match source.get(index + 1) {
          Some(b'=') => {
            index += 2;

            tokens.push(Token::ColonEquals {
              line: line,
              col: start_index - line_start_index,
            })
          }

          Some(b':') => {
            index += 2;

            tokens.push(Token::DoubleColon {
              line: line,
              col: start_index - line_start_index,
            })
          }

          _ => {
            index += 1;

            tokens.push(Token::Colon {
              line: line,
              col: start_index - line_start_index,
            })
          }
        },

        b'#' => {
          while index < length && source[index] != b'\n' {
            index += 1;
          }

          if self.preserve_comments {
            tokens.push(Token::Comment {
              line: line,
              col: start_index - line_start_index,
              value: &source[start_index + 1..index],
            })
          }
        }

        _ if is_identifier_start_char(byte) => {
          while index < length && is_identifier_char(source[index]) {
            index += 1;
          }

          tokens.push(Token::Identifier {
            line: line,
            col: start_index - line_start_index,
            value: &source[start_index..index],
          });

          continue;
        }

        _ if is_digit(byte) => {
          while index < length && is_digit(source[index]) {
            index += 1;
          }

          tokens.push(Token::Number {
            line: line,
            col: start_index - line_start_index,
            value: &source[start_index..index],
          });

          continue;
        }

        _ => {
          index += 1;

          tokens.push(Token::Unexpected {
            line: line,
            col: start_index - line_start_index,
          })
        }
      };
    }

    if interpolation_stack > 0 {
      return Err(TokenizeError{
        message: "Unclosed interpolation".to_owned(),
        line: 0, // TODO
        col: 0, // TODO
      });
    }

    if !string_stack.is_empty() {
      let (line_start, col_start, _) = string_stack.pop().unwrap();

      return Err(TokenizeError{
        message: "Unclosed string".to_owned(),
        line: line_start,
        col: col_start,
      });
    }

    Ok(tokens)
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use crate::assert_tokens_snapshot;
  use insta::assert_snapshot;

  assert_tokens_snapshot!(
    no_tokens,
    ""
  );

  assert_tokens_snapshot!(
    identifer_tokens,
    "hello world"
  );

  assert_tokens_snapshot!(
    number_tokens,
    "hello 1 47 wow"
  );

  assert_tokens_snapshot!(
    comment_tokens,
    "# o #nice\n# hello\ntest #same-line"
  );

  assert_tokens_snapshot!(
    symbol_tokens,
    "{ . } ( , ) : [ :: ] := = => ->"
  );

  assert_tokens_snapshot!(
    unexpected_tokens,
    "(@$@)"
  );

  assert_tokens_snapshot!(
    string_tokens_without_interpolations,
    "\"hello\" \"\" \"world\""
  );

  assert_tokens_snapshot!(
    string_tokens_with_interpolations,
    "\"hello $(name)!\" nice \"$(str)\""
  );

  assert_tokens_snapshot!(
    string_tokens_with_nested_interpolations,
    "\"hello $(name \"inner $(o)\" wow)!\""
  );
}
