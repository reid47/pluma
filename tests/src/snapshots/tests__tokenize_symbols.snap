---
source: compiler/src/tokenizer.rs
expression: "{ . } ( , ) : [ :: ] := = => ->"
---
Ok(
    (
        [
            LeftBrace(
                0,
                1,
            ),
            Dot(
                2,
                3,
            ),
            RightBrace(
                4,
                5,
            ),
            LeftParen(
                6,
                7,
            ),
            Comma(
                8,
                9,
            ),
            RightParen(
                10,
                11,
            ),
            Colon(
                12,
                13,
            ),
            LeftBracket(
                14,
                15,
            ),
            DoubleColon(
                16,
                18,
            ),
            RightBracket(
                19,
                20,
            ),
            ColonEquals(
                21,
                23,
            ),
            Equals(
                24,
                25,
            ),
            DoubleArrow(
                26,
                28,
            ),
            Arrow(
                29,
                31,
            ),
        ],
        {},
    ),
)
